src/win/tty.c

```diff
@@ -937,30 +956,106 @@ int uv_tty_read_start(uv_tty_t* handle, uv_alloc_cb alloc_cb,
 
 
 int uv_tty_read_stop(uv_tty_t* handle) {
+  INPUT_RECORD record;
+  DWORD written, err;
+
   handle->flags &= ~UV_HANDLE_READING;
   DECREASE_ACTIVE_COUNT(handle->loop, handle);
 
-  /* Cancel raw read */
-  if ((handle->flags & UV_HANDLE_READ_PENDING) &&
-      (handle->flags & UV_HANDLE_TTY_RAW)) {
+  if (!(handle->flags & UV_HANDLE_READ_PENDING))
+    return 0;
+
+  if (handle->flags & UV_HANDLE_TTY_RAW) {
+    /* Cancel raw read */
     /* Write some bullshit event to force the console wait to return. */
-    INPUT_RECORD record;
-    DWORD written;
     memset(&record, 0, sizeof record);
     if (!WriteConsoleInputW(handle->handle, &record, 1, &written)) {
       return GetLastError();
     }
+  } else if (!(handle->flags & UV_HANDLE_CANCELLATION_PENDING)) {
+    /* Cancel line-buffered read if not already pending */
+    err = uv__cancel_read_console(handle);
+    if (err)
+      return err;
+
+    handle->flags |= UV_HANDLE_CANCELLATION_PENDING;
+  }
+
+  return 0;
+}
+
+static int uv__cancel_read_console(uv_tty_t* handle) {
+  HANDLE active_screen_buffer = INVALID_HANDLE_VALUE;
+  CONSOLE_SCREEN_BUFFER_INFO info;
+  BOOL restore_screen_state;
+  INPUT_RECORD record;
+  DWORD written;
+  DWORD err = 0;
+
+  EnterCriticalSection(&uv__read_console_trap_lock);
+  if (!uv__read_console_started) {
+    /* We have managed to trap the other thread before ReadConsole is called */
+    uv__read_console_trap = TRUE;
+    LeaveCriticalSection(&uv__read_console_trap_lock);
+    return 0;
   }
+  LeaveCriticalSection(&uv__read_console_trap_lock);
+
+  /* If ReadConsole already returned (the user pressed ENTER),
+     there is nothing to do */
+  if (WaitForSingleObject(uv__read_console_done_event, 0) == WAIT_OBJECT_0)
```

Comment:
What if this thread gets to WaitForSingleObject before the 'reading' thread gets to ResetEvent ([there](https://github.com/janeasystems/libuv/blob/98548a05d5869ce31385efb3c70744ca1771a860/src/win/tty.c#L438))?